### PR TITLE
Customized for TEE&Occlum

### PR DESCRIPTION
**Pull request by NKcqx**
## Why are these changes needed?
In TEE(Trusted Execution Environment), there are some security-related restriction in OS level, like: No /dev/shm, `fork` system call is banned, etc., in which running a normal application may encounter problems like crash, hanging...
Therefore, some customized patches are needed in order to run Ray inside TEE.

1. `fork` -> `vfork`& `posix_spawn`;
2. replace sharing memory by `mmap` from `/dev/shm` by directly passing physical memory address since Occlum allows cross process memory accessing;
3. remove operations that may access `/proc/stat` since occlum only support limited fields of processes in the file;
4. disable memory monitor (temporally workaround)
5. optional compiling & running

 
## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: .bazelrc [hunk 1]
```diff
@@ -198,3 +198,7 @@ try-import %workspace%/.llvm-local.bazelrc
 build:macos --sandbox_block_path=/usr/local/
 #This option controls whether javac checks for missing direct dependencies.
 build --strict_java_deps=off
+
+# Running Ray inside TEE environment requires special compilation, normally
+# keep this flag to false unless you know what you are doing
+build --define RAY_IN_TEE=false
```
Comment (jovany-wang) [inline .bazelrc:204]: Not sure if we should use word `occlum` or `TEE`.

File: python/ray/_private/node.py [hunk 2]
```diff
@@ -305,6 +305,7 @@ def __init__(
 
         if not connect_only:
             self.start_ray_processes()
+            print("start_ray_processes end")
```
Comment (jovany-wang) [inline python/ray/_private/node.py:308]: Remove it.

File: python/ray/_private/node.py [hunk 3]
```diff
@@ -941,6 +942,7 @@ def start_raylet(
                 valgrind profiler.
         """
         stdout_file, stderr_file = self.get_log_file_handles("raylet", unique=True)
+        print("calling start_raylet")
```
Comment (jovany-wang) [inline python/ray/_private/node.py:945]: remove it.

File: python/ray/_private/node.py [hunk 4]
```diff
@@ -1058,28 +1061,36 @@ def start_head_processes(self):
         assert self._gcs_client is not None
         self._write_cluster_info_to_kv()
 
-        if not self._ray_params.no_monitor:
-            self.start_monitor()
+        if not ray_in_tee():
+            # Workaround because these params are not available in ray.init().
+            # Should be reverted once we start Ray with `ray start` CLI.
```
Comment (jovany-wang) [inline python/ray/_private/node.py:1066]: Why these should be reverted when we use `ray start`?

**Review by jovany-wang — COMMENTED** [review]
There are many verbose print messages, please remove them.

File: python/setup.py [hunk 5]
```diff
@@ -20,6 +20,8 @@
 
 logger = logging.getLogger(__name__)
 
+# A quick workaround to make sure we build Ray in Python 3.8
+# SUPPORTED_PYTHONS = [(3, 8)] # Occlum test only
 SUPPORTED_PYTHONS = [(3, 6), (3, 7), (3, 8), (3, 9), (3, 10)]
```
Comment (jovany-wang) [inline python/setup.py:25]: Do we need to change this line?

File: src/ray/boost_test.cc [hunk 6]
```diff
@@ -0,0 +1,34 @@
+#include <boost/asio.hpp>
```
Comment (jovany-wang) [inline src/ray/boost_test.cc:1]: Is this file necessary?

File: src/ray/common/client_connection.cc [hunk 7]
```diff
@@ -44,9 +44,11 @@ Status ConnectSocketRetry(local_stream_socket &socket,
   }
   boost::system::error_code ec;
   for (int num_attempts = 0; num_attempts < num_retries; ++num_attempts) {
+    RAY_LOG(INFO) << "ConnectSocketRetry begin " << endpoint << " " << num_attempts;
```
Comment (jovany-wang) [inline src/ray/common/client_connection.cc:47]: Remove these logs.

**Comment by jovany-wang** [comment]
Is it possible to add a unit test for running ray in Occlum?

**Comment by jovany-wang** [comment]
@zhouaihui @ian-huu @miluoalbert CC

**Comment by jovany-wang** [comment]
Let us build the `ENABLE_TEE=true` to occlum image by default? so that if it's a occlum image, that env works by default, without any manually exporting.

File: bazel/ray.bzl [hunk 8]
```diff
@@ -19,6 +19,10 @@ COPTS_WITHOUT_LOG = select({
     ],
     "//conditions:default": [
     ],
+}) + select({
+    "//:ray-in-tee": ["-DRAY_IN_TEE=1"],
```
[the remark this answers is not in the document]
Comment (NKcqx) [reply bazel/ray.bzl:23]: This is just a config flag, it only works when the build option `--define RAY_IN_TEE` is explicitly given.
So, this macro setting won't take the effect by default.

File: python/ray/_private/node.py (hunk 4 above)
Comment (NKcqx) [reply python/ray/_private/node.py:1063]: > Is it possible to add a unit test for running ray in Occlum?

Data transmission is done by passing a physical mem address in Occlum, this behavior will incur crashes in many stages, e.g. `ray.init`, outside the Occlum container.
Therefore, no particular interesting tests.

**Comment by NKcqx** [comment]
> Is it possible to add a unit test for running ray in Occlum?

Data transmission is done by passing a physical mem address in Occlum, this behavior will incur crashes in many stages, e.g. ray.init, outside the Occlum container.
Therefore, no particular interesting tests.
